File: app/src/main/java/com/matchthree/game/engine/LegalMoveDetector.kt
package com.matchthree.game.engine

import com.matchthree.game.model.Board
import com.matchthree.game.model.Gem
import com.matchthree.game.model.Position

/**
 * Counts swaps that would create a match. A swap is legal iff it is orthogonal
 * adjacency between two present gems and the swapped board contains a match.
 */
object LegalMoveDetector {

    fun legalMoveCount(board: Board): Int {
        var count = 0
        for (row in 0 until board.height) {
            for (col in 0 until board.width) {
                val a = Position(row, col)
                if (board.gemAt(a) == null) continue
                // Test each orthogonal neighbor once (right and down).
                val neighbors = listOf(
                    Position(row, col + 1),
                    Position(row + 1, col),
                )
                for (b in neighbors) {
                    if (board.isInside(b) && board.gemAt(b) != null) {
                        val swapped = board.withSwapped(a, b)
                        val matchLegal = MatchDetector.findMatches(swapped).isNotEmpty()
                        val specialLegal = SpecialRules.swapContactLegal(swapped.gemAt(a), swapped.gemAt(b))
                        if (matchLegal || specialLegal) count++
                    }
                }
            }
        }
        return count
    }

    fun hasLegalMove(board: Board): Boolean = legalMoveCount(board) > 0
}

File: Engine/Rules/LegalMoveDetector.cs
using MatchThree.Engine.Model;

namespace MatchThree.Engine.Rules;

public static class LegalMoveDetector
{
    public static int LegalMoveCount(Board board)
    {
        var count = 0;
        for (var row = 0; row < board.Height; row++)
        {
            for (var col = 0; col < board.Width; col++)
            {
                var a = new Position(row, col);
                if (board.GemAt(a) is null) continue;

                var neighbors = new[] { new Position(row, col + 1), new Position(row + 1, col) };
                foreach (var b in neighbors)
                {
                    if (board.IsInside(b) && board.GemAt(b) is not null)
                    {
                        var swapped = board.WithSwapped(a, b);
                        var matchLegal = MatchDetector.FindMatches(swapped).Count > 0;
                        var specialLegal = SpecialRules.SwapContactLegal(swapped.GemAt(a), swapped.GemAt(b));
                        if (matchLegal || specialLegal) count++;
                    }
                }
            }
        }
        return count;
    }

    public static bool HasLegalMove(Board board) => LegalMoveCount(board) > 0;
}